### ISSUE TITLE
Conflict errors on Azure after retrying `insert_entity` (because original insert really succeeded)

### ISSUE THREAD

**Issue by adamnovak** [BD2KGenomics/toil#579]
I've noticed Toil producing an occasional Conflict error when trying to write things to the Azure JobStore (with my #578 patch) when creating a new job:

```
WARNING:toil.leader:9eca3f3b_fafc_4b1a_b060_f16a72308246:       INFO:toil.jobStores.azureJobStore:Exception WindowsAzureError passed 
predicate, trying again
WARNING:toil.leader:9eca3f3b_fafc_4b1a_b060_f16a72308246:       Traceback (most recent call last):
WARNING:toil.leader:9eca3f3b_fafc_4b1a_b060_f16a72308246:         File "/usr/local/lib/python2.7/dist-packages/toil/worker.py", line 
282, in main
WARNING:toil.leader:9eca3f3b_fafc_4b1a_b060_f16a72308246:           fileStore=fileStore)
WARNING:toil.leader:9eca3f3b_fafc_4b1a_b060_f16a72308246:         File "/usr/local/lib/python2.7/dist-packages/toil/job.py", line 128
6, in _execute
WARNING:toil.leader:9eca3f3b_fafc_4b1a_b060_f16a72308246:           self._serialiseJobGraph(jobWrapper, jobStore, returnValues, False
)
WARNING:toil.leader:9eca3f3b_fafc_4b1a_b060_f16a72308246:         File "/usr/local/lib/python2.7/dist-packages/toil/job.py", line 121
4, in _serialiseJobGraph
WARNING:toil.leader:9eca3f3b_fafc_4b1a_b060_f16a72308246:           jobsToJobWrappers = self._makeJobWrappers(jobWrapper, jobStore)
WARNING:toil.leader:9eca3f3b_fafc_4b1a_b060_f16a72308246:         File "/usr/local/lib/python2.7/dist-packages/toil/job.py", line 112
9, in _makeJobWrappers
WARNING:toil.leader:9eca3f3b_fafc_4b1a_b060_f16a72308246:           successor._makeJobWrappers2(jobStore, jobsToJobWrappers), successors)
WARNING:toil.leader:9eca3f3b_fafc_4b1a_b060_f16a72308246:         File "/usr/local/lib/python2.7/dist-packages/toil/job.py", line 1129, in <lambda>
WARNING:toil.leader:9eca3f3b_fafc_4b1a_b060_f16a72308246:           successor._makeJobWrappers2(jobStore, jobsToJobWrappers), successors)
WARNING:toil.leader:9eca3f3b_fafc_4b1a_b060_f16a72308246:         File "/usr/local/lib/python2.7/dist-packages/toil/job.py", line 1141, in _makeJobWrappers2
WARNING:toil.leader:9eca3f3b_fafc_4b1a_b060_f16a72308246:           successor._makeJobWrappers2(jobStore, jobsToJobWrappers), successors)
WARNING:toil.leader:9eca3f3b_fafc_4b1a_b060_f16a72308246:         File "/usr/local/lib/python2.7/dist-packages/toil/job.py", line 1141, in <lambda>
WARNING:toil.leader:9eca3f3b_fafc_4b1a_b060_f16a72308246:           successor._makeJobWrappers2(jobStore, jobsToJobWrappers), successors)
WARNING:toil.leader:9eca3f3b_fafc_4b1a_b060_f16a72308246:         File "/usr/local/lib/python2.7/dist-packages/toil/job.py", line 1136, in _makeJobWrappers2
WARNING:toil.leader:9eca3f3b_fafc_4b1a_b060_f16a72308246:           jobWrapper = self._createEmptyJobForJob(jobStore, predecessorNumber=len(self._directPredecessors))
WARNING:toil.leader:9eca3f3b_fafc_4b1a_b060_f16a72308246:         File "/usr/local/lib/python2.7/dist-packages/toil/job.py", line 1099, in _createEmptyJobForJob
WARNING:toil.leader:9eca3f3b_fafc_4b1a_b060_f16a72308246:           return jobStore.create(command=command, predecessorNumber=predecessorNumber, **requirements)
WARNING:toil.leader:9eca3f3b_fafc_4b1a_b060_f16a72308246:         File "/usr/local/lib/python2.7/dist-packages/toil/jobStores/azureJobStore.py", line 137, in create
WARNING:toil.leader:9eca3f3b_fafc_4b1a_b060_f16a72308246:           self.jobItems.insert_entity(entity=entity)
WARNING:toil.leader:9eca3f3b_fafc_4b1a_b060_f16a72308246:         File "/usr/local/lib/python2.7/dist-packages/toil/jobStores/azureJobStore.py", line 515, in f
WARNING:toil.leader:9eca3f3b_fafc_4b1a_b060_f16a72308246:           return function(**kwargs)
WARNING:toil.leader:9eca3f3b_fafc_4b1a_b060_f16a72308246:         File "/usr/local/lib/python2.7/dist-packages/azure/storage/tableservice.py", line 431, in insert_entity
WARNING:toil.leader:9eca3f3b_fafc_4b1a_b060_f16a72308246:           response = self._perform_request(request)
WARNING:toil.leader:9eca3f3b_fafc_4b1a_b060_f16a72308246:         File "/usr/local/lib/python2.7/dist-packages/azure/storage/storageclient.py", line 179, in _perform_request
WARNING:toil.leader:9eca3f3b_fafc_4b1a_b060_f16a72308246:           _storage_error_handler(ex)
WARNING:toil.leader:9eca3f3b_fafc_4b1a_b060_f16a72308246:         File "/usr/local/lib/python2.7/dist-packages/azure/storage/__init__.py", line 1174, in _storage_error_handler
WARNING:toil.leader:9eca3f3b_fafc_4b1a_b060_f16a72308246:           return _general_error_handler(http_error)
WARNING:toil.leader:9eca3f3b_fafc_4b1a_b060_f16a72308246:         File "/usr/local/lib/python2.7/dist-packages/azure/__init__.py", line 543, in _general_error_handler
WARNING:toil.leader:9eca3f3b_fafc_4b1a_b060_f16a72308246:           _ERROR_CONFLICT.format(str(http_error)))
WARNING:toil.leader:9eca3f3b_fafc_4b1a_b060_f16a72308246:       WindowsAzureConflictError: Conflict (Conflict)
WARNING:toil.leader:9eca3f3b_fafc_4b1a_b060_f16a72308246:       Exiting the worker because of a failed jobWrapper on host c1agent1
WARNING:toil.leader:9eca3f3b_fafc_4b1a_b060_f16a72308246:       ERROR:toil.worker:Exiting the worker because of a failed jobWrapper on host c1agent1
WARNING:toil.leader:9eca3f3b_fafc_4b1a_b060_f16a72308246:       WARNING:toil.jobWrapper:Due to failure we are reducing the remaining retry count of job 9eca3f3b_fafc_4b1a_b060_f16a72308246 to 0
WARNING:toil.leader:Job: 9eca3f3b_fafc_4b1a_b060_f16a72308246 is completely failed
```

And previously, when associating a file with a job:

```
WARNING:toil.leader:45cc36d2_9bfe_477f_b085_52135498ea7c:   INFO:toil.jobStores.azureJobStore:Got a retriable exception SSLError, trying again
WARNING:toil.leader:45cc36d2_9bfe_477f_b085_52135498ea7c:   Traceback (most recent call last):
WARNING:toil.leader:45cc36d2_9bfe_477f_b085_52135498ea7c:     File "/usr/local/lib/python2.7/dist-packages/toil/worker.py", line 282, in main
WARNING:toil.leader:45cc36d2_9bfe_477f_b085_52135498ea7c:       fileStore=fileStore)
WARNING:toil.leader:45cc36d2_9bfe_477f_b085_52135498ea7c:     File "/usr/local/lib/python2.7/dist-packages/toil/job.py", line 1282, in _execute
WARNING:toil.leader:45cc36d2_9bfe_477f_b085_52135498ea7c:       self._serialiseJobGraph(jobWrapper, jobStore, returnValues, False)
WARNING:toil.leader:45cc36d2_9bfe_477f_b085_52135498ea7c:     File "/usr/local/lib/python2.7/dist-packages/toil/job.py", line 1233, in _serialiseJobGraph
WARNING:toil.leader:45cc36d2_9bfe_477f_b085_52135498ea7c:       job._serialiseJob(jobStore, jobsToJobWrappers, jobWrapper)
WARNING:toil.leader:45cc36d2_9bfe_477f_b085_52135498ea7c:     File "/usr/local/lib/python2.7/dist-packages/toil/job.py", line 1186, in _serialiseJob
WARNING:toil.leader:45cc36d2_9bfe_477f_b085_52135498ea7c:       cPickle.dump(self, fileHandle, cPickle.HIGHEST_PROTOCOL)
WARNING:toil.leader:45cc36d2_9bfe_477f_b085_52135498ea7c:     File "/usr/lib/python2.7/contextlib.py", line 24, in __exit__
WARNING:toil.leader:45cc36d2_9bfe_477f_b085_52135498ea7c:       self.gen.next()
WARNING:toil.leader:45cc36d2_9bfe_477f_b085_52135498ea7c:     File "/usr/local/lib/python2.7/dist-packages/toil/jobStores/azureJobStore.py", line 228, in writeFileStream
WARNING:toil.leader:45cc36d2_9bfe_477f_b085_52135498ea7c:       self._associateFileWithJob(jobStoreFileID, jobStoreID)
WARNING:toil.leader:45cc36d2_9bfe_477f_b085_52135498ea7c:     File "/usr/local/lib/python2.7/dist-packages/toil/jobStores/azureJobStore.py", line 324, in _associateFileWithJob
WARNING:toil.leader:45cc36d2_9bfe_477f_b085_52135498ea7c:       'RowKey': jobStoreFileID})
WARNING:toil.leader:45cc36d2_9bfe_477f_b085_52135498ea7c:     File "/usr/local/lib/python2.7/dist-packages/toil/jobStores/azureJobStore.py", line 517, in f
WARNING:toil.leader:45cc36d2_9bfe_477f_b085_52135498ea7c:       return function(**kwargs)
WARNING:toil.leader:45cc36d2_9bfe_477f_b085_52135498ea7c:     File "/usr/local/lib/python2.7/dist-packages/azure/storage/tableservice.py", line 431, in insert_entity
WARNING:toil.leader:45cc36d2_9bfe_477f_b085_52135498ea7c:       response = self._perform_request(request)
WARNING:toil.leader:45cc36d2_9bfe_477f_b085_52135498ea7c:     File "/usr/local/lib/python2.7/dist-packages/azure/storage/storageclient.py", line 179, in _perform_request
WARNING:toil.leader:45cc36d2_9bfe_477f_b085_52135498ea7c:       _storage_error_handler(ex)
WARNING:toil.leader:45cc36d2_9bfe_477f_b085_52135498ea7c:     File "/usr/local/lib/python2.7/dist-packages/azure/storage/__init__.py", line 1174, in _storage_error_handler
WARNING:toil.leader:45cc36d2_9bfe_477f_b085_52135498ea7c:       return _general_error_handler(http_error)
WARNING:toil.leader:45cc36d2_9bfe_477f_b085_52135498ea7c:     File "/usr/local/lib/python2.7/dist-packages/azure/__init__.py", line 543, in _general_error_handler
WARNING:toil.leader:45cc36d2_9bfe_477f_b085_52135498ea7c:       _ERROR_CONFLICT.format(str(http_error)))
WARNING:toil.leader:45cc36d2_9bfe_477f_b085_52135498ea7c:   WindowsAzureConflictError: Conflict (Conflict)
WARNING:toil.leader:45cc36d2_9bfe_477f_b085_52135498ea7c:   Exiting the worker because of a failed jobWrapper on host c1agent33
WARNING:toil.leader:45cc36d2_9bfe_477f_b085_52135498ea7c:   ERROR:toil.worker:Exiting the worker because of a failed jobWrapper on host c1agent33
WARNING:toil.leader:45cc36d2_9bfe_477f_b085_52135498ea7c:   Exception in thread Thread-4:
WARNING:toil.leader:45cc36d2_9bfe_477f_b085_52135498ea7c:   Traceback (most recent call last):
WARNING:toil.leader:45cc36d2_9bfe_477f_b085_52135498ea7c:     File "/usr/lib/python2.7/threading.py", line 810, in __bootstrap_inner
WARNING:toil.leader:45cc36d2_9bfe_477f_b085_52135498ea7c:       self.run()
WARNING:toil.leader:45cc36d2_9bfe_477f_b085_52135498ea7c:     File "/usr/lib/python2.7/threading.py", line 763, in run
WARNING:toil.leader:45cc36d2_9bfe_477f_b085_52135498ea7c:       self.__target(*self.__args, **self.__kwargs)
WARNING:toil.leader:45cc36d2_9bfe_477f_b085_52135498ea7c:     File "/usr/local/lib/python2.7/dist-packages/toil/job.py", line 487, in asyncWrite
WARNING:toil.leader:45cc36d2_9bfe_477f_b085_52135498ea7c:       raise RuntimeError("The termination flag is set, exiting")
WARNING:toil.leader:45cc36d2_9bfe_477f_b085_52135498ea7c:   RuntimeError: The termination flag is set, exiting
WARNING:toil.leader:45cc36d2_9bfe_477f_b085_52135498ea7c:   
WARNING:toil.leader:45cc36d2_9bfe_477f_b085_52135498ea7c:   Exception in thread Thread-5:
WARNING:toil.leader:45cc36d2_9bfe_477f_b085_52135498ea7c:   Traceback (most recent call last):
WARNING:toil.leader:45cc36d2_9bfe_477f_b085_52135498ea7c:     File "/usr/lib/python2.7/threading.py", line 810, in __bootstrap_inner
WARNING:toil.leader:45cc36d2_9bfe_477f_b085_52135498ea7c:       self.run()
WARNING:toil.leader:45cc36d2_9bfe_477f_b085_52135498ea7c:     File "/usr/lib/python2.7/threading.py", line 763, in run
WARNING:toil.leader:45cc36d2_9bfe_477f_b085_52135498ea7c:       self.__target(*self.__args, **self.__kwargs)
WARNING:toil.leader:45cc36d2_9bfe_477f_b085_52135498ea7c:     File "/usr/local/lib/python2.7/dist-packages/toil/job.py", line 487, in asyncWrite
WARNING:toil.leader:45cc36d2_9bfe_477f_b085_52135498ea7c:       raise RuntimeError("The termination flag is set, exiting")
WARNING:toil.leader:45cc36d2_9bfe_477f_b085_52135498ea7c:   RuntimeError: The termination flag is set, exiting
WARNING:toil.leader:45cc36d2_9bfe_477f_b085_52135498ea7c:   
WARNING:toil.leader:45cc36d2_9bfe_477f_b085_52135498ea7c:   WARNING:toil.jobWrapper:Due to failure we are reducing the remaining retry count of job 45cc36d2_9bfe_477f_b085_52135498ea7c to 0
```

Both of these are instances of `insert_entity` throwing a retriable error, being retried, and failing the retry attempt due to a conflict. I think that the original insert is taking even though it returned an error, and thus the repeated attempt to insert the same thing (as opposed to updating it) fails.

Inserts should be retried as updates, or conflicts on retries of inserts should be taken as evidence of success (assuming Azure can't be made to accidentally make a partial insert).

**Comment by hannes-ucsc** [comment]
The simple fix, which is to substitute `insert_entity` with `insert_or_replace_entity` produces very surprising test failures, mostly with deletions failing. Need more time to look into this.

**Comment by hannes-ucsc** [comment]
Consider 5b33d1cbf2d452b3bb52545ee34bb3152b0be5c6 as part of a solution.

**Comment by hannes-ucsc** [comment]
http://jenkins.cgcloud.info/job/toil-pull-requests/1402/testReport/junit/src.toil.test.src.jobCacheTest/AzureJobStoreCacheTest/testDeleteLocalFile/

**Comment by hannes-ucsc** [comment]
http://jenkins.cgcloud.info/job/toil-pull-requests/1406/testReport/junit/src.toil.test.sort.sortTest/SortTest/testAzureSingle/

**Comment by DailyDreaming** [comment]
Addressed in #1912.
